contracts/TimeMedianDataFeedOracle.sol

```diff
@@ -0,0 +1,137 @@
+
+  function medianizeByTimeframe(uint startDate, uint endDate)
+  public
+  returns (bytes32 medianValue) {
+    require(startDate < endDate, 'startDate must be less than endDate');
+    uint endIndex;
+    uint startIndex;
+
+    // find end date within dates array
+    for (uint i = dates.length - 1; i > 0 && endIndex == 0; i--) {
+      if (dates[i] <= endDate) {
+        endIndex = i; // ONLY dates before the end date
+      }
+    }
+
+    // require endIndex is more than 0
+    require(endIndex > 0, 'dates outside of range of datafeed results');
+    require(dates[endIndex] >= startDate, 'dates outside of range of datafeed results');
+
+    // find start date within dates array
+    for (uint j = endIndex; j >= 0 && startIndex == 0; j--) {
+      if (dates[j] < startDate) {
+        startIndex = j + 1; // no dates before the startDate
+      }
+    }
+
+    uint[] memory selectedDates = _createPartialArray(startIndex, endIndex);
+    uint[] memory partitionedDates = partitionDates(selectedDates);
+    medianValue = _medianizeByDates(partitionedDates);
+    emit MedianizedByTimeframe(medianValue, startDate, endDate);
+  }
+
+  function medianizeByDates(uint[] orderedDates)
+  public
+  returns (bytes32 medianValue) {
+    for (uint i = 0; i < orderedDates.length; i++) {
+      uint date = orderedDates[i];
+      require(isResultSetFor(date), "Date not set.");
+      require(!dateAlreadyAccountedFor[date], 'Date cannot be a duplicate');
+      dateAlreadyAccountedFor[orderedDates[i]] = true;
+      if (i != orderedDates.length - 1) {
+        require(uint(results[date]) <= uint(results[orderedDates[i+1]]), "The dates are not sorted by result.");
+      }
+    }
+
+    // reset dataFeedAlreadyRecorded
+    for(uint j=0; j < orderedDates.length; j++) {
+      dateAlreadyAccountedFor[orderedDates[j]] = false;
+    }
+
+    medianValue = _medianizeByDates(orderedDates);
+    emit MedianizedByOrderedDates(medianValue, orderedDates);
+  }
+
+  // Private Functions
+
+  event MiddleIndex(uint middleIndex);
+
+  function _medianizeByDates(uint[] orderedDates)
+  private
+  returns (bytes32 medianValue) {
+    uint middleIndex = orderedDates.length / 2;
+    uint middleDate  = orderedDates[middleIndex];
+    medianValue = results[middleDate];
+  }
+
+  function partitionDates(uint[] selectedDates)
+  private
+  returns (uint[]) {
+    // To minimize complexity, return the higher of the two middle checkpoints in even-sized arrays instead of the average.
+    uint k = selectedDates.length.div(2);
+    uint left = 0;
+    uint right = selectedDates.length.sub(1);
+
+    while (left < right) {
+        uint pivotIndex = left.add(right).div(2);
+        uint pivotValue = selectedDates[pivotIndex];
+
+        (selectedDates[pivotIndex], selectedDates[right]) = (selectedDates[right], selectedDates[pivotIndex]);
+        uint storeIndex = left;
+        for (uint i = left; i < right; i++) {
+            if (_isLessThan(selectedDates[i], pivotValue)) {
```

Comment:
should this be checking `results` mapping by the date, instead of the date value itself?

Reply:
nevermind, I see the check now